Pod/Classes/BONTag.m

```diff
@@ -0,0 +1,235 @@
+//
+//  BONTag.m
+//  Pods
+//
+//  Created by Nora Trapp on 5/11/16.
+//
+//
+
+#import "BONTag_Private.h"
+#import "BONMot.h"
+
+static const NSString *kBONTagDefaultStartPrefix = @"<";
+static const NSString *kBONTagDefaultStartSuffix = @">";
+static const NSString *kBONTagDefaultEndPrefix = @"</";
+static const NSString *kBONTagDefaultEndSuffix = @">";
+static const NSString *kBONTagDefaultEscapeString = @"\\";
+
+@interface BONTag ()
+
+@property (copy, nonatomic) NSString *startTag;
+@property (copy, nonatomic) NSString *endTag;
+@property (copy, nonatomic) NSString *escapeString;
+@property (copy, nonatomic) id<BONTextable> textable;
+
+@end
+
+@implementation BONTag
+
+- (instancetype)initWithStartTag:(NSString *)startTag endTag:(NSString *)endTag escapeString:(NSString *)escapeString textable:(id<BONTextable>)textable
+{
+    self = [super init];
+    if (self) {
+        self.startTag = startTag;
+        self.endTag = endTag;
+        self.escapeString = escapeString;
+        self.textable = textable;
+        self.ranges = [NSMutableArray array];
+    }
+
+    return self;
+}
+
+- (instancetype)initWithTag:(NSString *)tag textable:(id<BONTextable>)textable
+{
+    NSString *startTag = [NSString stringWithFormat:@"%@%@%@", kBONTagDefaultStartPrefix, tag, kBONTagDefaultStartSuffix];
+    NSString *endTag = [NSString stringWithFormat:@"%@%@%@", kBONTagDefaultEndPrefix, tag, kBONTagDefaultEndSuffix];
+
+    return [self initWithStartTag:startTag endTag:endTag escapeString:(NSString *)kBONTagDefaultEscapeString textable:textable];
+}
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    __typeof(self) tag = [[self.class alloc] init];
+
+    tag.startTag = self.startTag;
+    tag.endTag = self.endTag;
+    tag.escapeString = self.escapeString;
+    tag.textable = self.textable;
+    tag.ranges = self.ranges;
+
+    return tag;
+}
+
+- (NSString *)description
+{
+    NSString *description = [NSString stringWithFormat:@"<%@: %p, startTag: %@, endTag: %@, textable: <%@: %p>>", NSStringFromClass(self.class), self, self.startTag, self.endTag, NSStringFromClass(self.textable.class), self.textable];
+
+    return description;
+}
+
+#pragma mark - Tag Matching
+
++ (NSArray *)escapedRangesInString:(NSString **)string withTags:(NSArray *)tags stripEscapeCharacters:(BOOL)stripEscapeCharacters
+{
+    NSMutableArray *escapedRanges = [NSMutableArray array];
+
+    NSString *theString = *string;
```

Comment:
#req even though this is declared as non-null, in Obj-C, it could still come in as null. Prefer asserting explicitly rather than randomly crashing.